### PR TITLE
Add LDAP authenticator variant using `ldap3` package

### PR DESCRIPTION
**Pull request by mahendrapaipuri**
Currently, LDAP authenticator uses `python-ldap` package which needs OpenLDAP client libraries on the host machine for working properly. If there are no OpenLDAP client libraries found on the host, it will bring in pre-compiled libraries and build the package against them. I have noticed some compatibility issues in this case.

This PR adds a new variant of the LDAP authenticator based on [ldap3](https://ldap3.readthedocs.io/en/latest/index.html) which is pure Python implementation of OpenLDAP client libraries. This package is more robust than `python-ldap` and works out-of-the-box.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. I left the boiler plate code in the `ldap_ad.py` file so the maintainers can test the implementation.
  
## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by bernt-matthias** [comment]
If needed / wanted I could test this (after Nov 9).

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 1]
```diff
@@ -12,14 +20,20 @@
     string_as_bool,
     unicodify,
 )
-from ..providers import AuthProvider
+from galaxy.auth.providers import AuthProvider
```
Comment (nsoranzo) [inline lib/galaxy/auth/providers/ldap_ad.py]:
```suggestion
from . import AuthProvider
```

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 2]
```diff
@@ -2,6 +2,14 @@
 Created on 15/07/2014
 
 @author: Andrew Robinson
+
+Modification on 24/10/2022
+
+Addition of LDAP auth provider using ldap3 module. Current LDAP auth provider uses python-ldap module which
+has external dependencies like openldap client libs. ldap3 is pure Python implementation of OpenLDAP client library 
+and is more robust than python-ldap.
```
Comment (nsoranzo) [inline lib/galaxy/auth/providers/ldap_ad.py]:
```suggestion
has external dependencies like the openldap client libs. ldap3 is instead a pure Python LDAP v3 client library.
```

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 3]
```diff
@@ -2,6 +2,14 @@
 Created on 15/07/2014
 
 @author: Andrew Robinson
+
+Modification on 24/10/2022
+
+Addition of LDAP auth provider using ldap3 module. Current LDAP auth provider uses python-ldap module which
```
Comment (nsoranzo) [inline lib/galaxy/auth/providers/ldap_ad.py]:
```suggestion
Addition of the LDAP3 auth provider using the ldap3 library. The original LDAP auth provider uses the python-ldap library, which
```

**Comment by mahendrapaipuri** [comment]
@bernt-matthias I left the boiler plate to test the implementation. @nsoranzo Cheers for the review!

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 4]
```diff
@@ -282,11 +295,179 @@ def authenticate_user(self, user, password, options, request):
...
+            try:
+                whoami = conn.extend.standard.who_am_i()
+                # Unbind connection
+                conn.unbind()
+            except Exception:
+                # The "Who am I?" extended operation is not supported by this LDAP server
+                pass
+            else:
+                if whoami is None:
+                    raise RuntimeError("LDAP3 authenticate: anonymous bind")
```
Comment (bernt-matthias) [inline lib/galaxy/auth/providers/ldap_ad.py:438]: Is this bad in general? (note: I have not so much an idea of LDAP. I'm just a user)

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 5]
```diff
@@ -282,11 +295,179 @@ def authenticate_user(self, user, password, options, request):
...
+        # LDAP over TLS bool
+        self.ldap_tls = None
+
+    def get_server(self, options, params):
+        # Get server URL
+        server_url = _get_subs(options, "server", params)
+        # Check if server URL has scheme
+        # If no scheme is provided, assume it to be ldap
+        if "ldap" not in server_url:
+            server_url = "ldap://{server_url}".format(server_url=server_url)
```
Comment (bernt-matthias) [inline lib/galaxy/auth/providers/ldap_ad.py]: maybe default to `ldaps`?

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 6]
```diff
@@ -282,11 +295,179 @@ def authenticate_user(self, user, password, options, request):
...
+        # Check if TLS is available
+        if server_url.startswith("ldaps://"):
+            self.ldap_tls = True
+        else:
+            self.ldap_tls = False
+        try:
+            # Get server address and port
+            server_address, server_port = server_url.split("//")[-1].split(":")
```
Comment (bernt-matthias) [inline lib/galaxy/auth/providers/ldap_ad.py]: Is there something in the python stdlib to split URIs?

File: lib/galaxy/dependencies/conditional-requirements.txt [hunk 7]
```diff
@@ -9,6 +9,7 @@ statsd
 azure-storage==0.32.0
 python-irodsclient==1.1.3
 python-ldap==3.4.0
+ldap3==2.9.1
```
Comment (bernt-matthias) [inline lib/galaxy/dependencies/conditional-requirements.txt:12]: Wondering if it makes sense to have both in this file?

**Comment by bernt-matthias** [comment]
I could test this successfully with our LDAP server. Just needed to do the following:

- in auth_conf: switch to `<type>ldap3</type>`
- switched python module: 

```
# activate venv
pip uninstall python-ldap
pip install ldap3
pip install ldap3==2.9.1
# deactivate venv
```

File: lib/galaxy/dependencies/conditional-requirements.txt (hunk 7 above)
Comment (nsoranzo) [reply lib/galaxy/dependencies/conditional-requirements.txt:12]: It doesn't hurt, but to trigger auto installation a `check_ldap3()` method similar to `check_python_ldap()` needs to be added to `lib/galaxy/dependencies/__init__.py` .

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 8]
```diff
@@ -282,11 +295,179 @@ def authenticate_user(self, user, password, options, request):
...
+                auto_bind=self.auto_bind,
+            )
+            # Use StartTLS if LDAP connection is not over TLS
+            if not self.ldap_tls:
+                conn.start_tls()
+            try:
+                whoami = conn.extend.standard.who_am_i()
+                # Unbind connection
+                conn.unbind()
+            except Exception:
```
Comment (nsoranzo) [inline lib/galaxy/auth/providers/ldap_ad.py]: Can you restrict this exception?

File: lib/galaxy/auth/providers/ldap_ad.py (hunk 4 above)
Comment (nsoranzo) [reply lib/galaxy/auth/providers/ldap_ad.py:438]: This mirrors the older ldap implementation and looks fine to me.

**Comment by bernt-matthias** [comment]
Wondering if we can improve docs. In particular the point that admins now have a choice. Is there a list of pros and cons? Can we make a suggestion? Maybe adapt auth_conf.xml.sample?

File: lib/galaxy/auth/providers/ldap_ad.py (hunk 4 above)
Comment (mahendrapaipuri) [reply lib/galaxy/auth/providers/ldap_ad.py:438]: Exactly!! I tried to not to change the authentication logic from the original implementation. The major difference between two implementations is setting up the connection object.

File: lib/galaxy/auth/providers/ldap_ad.py (hunk 5 above)
Comment (mahendrapaipuri) [reply lib/galaxy/auth/providers/ldap_ad.py]: Makes sense. I will change it!

File: lib/galaxy/auth/providers/ldap_ad.py (hunk 6 above)
Comment (mahendrapaipuri) [reply lib/galaxy/auth/providers/ldap_ad.py]: Not that I know. For HTTP, there is one. Not sure if it works for LDAP. Will check it. Cheers

File: lib/galaxy/dependencies/conditional-requirements.txt (hunk 7 above)
Comment (mahendrapaipuri) [reply lib/galaxy/dependencies/conditional-requirements.txt:12]: Cool. Not aware of `check_python_ldap()` method. I can add a method as suggested.

File: lib/galaxy/auth/providers/ldap_ad.py (hunk 8 above)
Comment (mahendrapaipuri) [reply lib/galaxy/auth/providers/ldap_ad.py]: Cheers. Nice find.  We need to catch [`LDAPExtensionError`](https://github.com/cannatag/ldap3/blob/78c449dded29dad54e4e68055f7290ad8689fbd1/test/testExtendedOperations.py#L57-L65). Will change it.

File: doc/source/admin/authentication.md [hunk 9]
```diff
@@ -30,6 +30,12 @@ Galaxy is distributed with a plugin-driven authentication framework for which th
 just one (and the default plugin). This framework can be used to allow Galaxy to delegate authentication to
 an LDAP server, an Active Directory server, or to PAM.
 
+Currently, we provide two variants of LDAP authenticator namely, `ldap` and `ldap3`. Both are identical implementations that use different Python modules for binding and making queries to a LDAP server. The authenticator
+`ldap` is based on [python-ldap](https://www.python-ldap.org/en/python-ldap-3.4.3/) module which is a wrapper around 
+OpenLDAP's client library `libldap`. As of writing, `python-ldap` does not provide pre-build wheel packages and hence, OpenLDAP client libraries are prerequisites for building `python-ldap`. 
+
+On the otherhand, `ldap3` is pure Python implementation of OpenLDAP client library and has no external dependencies. This package works out-of-the-box and we recommend to use it when OpenLDAP client libraries are available on the host.
```
Comment (bernt-matthias) [inline doc/source/admin/authentication.md]:
```suggestion
On the otherhand, `ldap3` is pure Python implementation of OpenLDAP client library and has no external dependencies. This package works out-of-the-box and we recommend to use it when OpenLDAP client libraries are not available on the host.
```

its "not available" or?

**Comment by mahendrapaipuri** [comment]
@bernt-matthias @nsoranzo Is there anything else to be done for this PR? Cheers!!

**Review by bernt-matthias — APPROVED** [review]
Fine from my side. @nsoranzo is there anything in the requirements that need to be changed?

File: doc/source/admin/authentication.md [hunk 10]
```diff
@@ -30,6 +30,12 @@ Galaxy is distributed with a plugin-driven authentication framework for which th
 just one (and the default plugin). This framework can be used to allow Galaxy to delegate authentication to
 an LDAP server, an Active Directory server, or to PAM.
 
+Currently, we provide two variants of LDAP authenticator namely, `ldap` and `ldap3`. Both are identical implementations that use different Python modules for binding and making queries to a LDAP server. The authenticator
```
Comment (nsoranzo) [inline doc/source/admin/authentication.md]:
```suggestion
Currently, we provide two variants of the LDAP authenticator namely, `ldap` and `ldap3`. Both are identical implementations that use different Python modules for binding and making queries to a LDAP server. The authenticator
```

File: doc/source/admin/authentication.md [hunk 11]
```diff
@@ -30,6 +30,12 @@ Galaxy is distributed with a plugin-driven authentication framework for which th
...
+`ldap` is based on [python-ldap](https://www.python-ldap.org/en/python-ldap-3.4.3/) module which is a wrapper around
```
Comment (nsoranzo) [inline doc/source/admin/authentication.md]:
```suggestion
`ldap` is based on the [python-ldap](https://www.python-ldap.org/) module which is a wrapper around the
```

File: doc/source/admin/authentication.md [hunk 12]
```diff
@@ -30,6 +30,12 @@ Galaxy is distributed with a plugin-driven authentication framework for which th
 just one (and the default plugin). This framework can be used to allow Galaxy to delegate authentication to
 an LDAP server, an Active Directory server, or to PAM.
 
+Currently, we provide two variants of LDAP authenticator namely, `ldap` and `ldap3`. Both are identical implementations that use different Python modules for binding and making queries to a LDAP server. The authenticator
+`ldap` is based on [python-ldap](https://www.python-ldap.org/en/python-ldap-3.4.3/) module which is a wrapper around 
+OpenLDAP's client library `libldap`. As of writing, `python-ldap` does not provide pre-build wheel packages and hence, OpenLDAP client libraries are prerequisites for building `python-ldap`.
```
Comment (nsoranzo) [inline doc/source/admin/authentication.md]:
```suggestion
OpenLDAP's client library `libldap`. Currently `python-ldap` does not provide pre-built Python wheel packages, hence the OpenLDAP client libraries are needed to install `python-ldap` on the Galaxy server. 
```

File: doc/source/admin/authentication.md [hunk 13]
```diff
@@ -30,6 +30,12 @@ Galaxy is distributed with a plugin-driven authentication framework for which th
 just one (and the default plugin). This framework can be used to allow Galaxy to delegate authentication to
 an LDAP server, an Active Directory server, or to PAM.
 
+Currently, we provide two variants of LDAP authenticator namely, `ldap` and `ldap3`. Both are identical implementations that use different Python modules for binding and making queries to a LDAP server. The authenticator
+`ldap` is based on [python-ldap](https://www.python-ldap.org/en/python-ldap-3.4.3/) module which is a wrapper around 
+OpenLDAP's client library `libldap`. As of writing, `python-ldap` does not provide pre-build wheel packages and hence, OpenLDAP client libraries are prerequisites for building `python-ldap`. 
+
+On the otherhand, `ldap3` is pure Python implementation of OpenLDAP client library and has no external dependencies. This package works out-of-the-box and we recommend to use it when OpenLDAP client libraries are not available on the host.
```
Comment (nsoranzo) [inline doc/source/admin/authentication.md]:
```suggestion
On the otherhand, `ldap3` is a pure-Python implementation of the OpenLDAP client library and has no external dependencies. This package can be installed out-of-the-box, so we recommend to use it when the OpenLDAP client libraries are not available on the Galaxy server.
```

**Review by nsoranzo — APPROVED** [review]

**Comment by nsoranzo** [comment]
Thanks @mahendrapaipuri !

**Pull request merged**